reuse.go

```diff
@@ -0,0 +1,140 @@
+package libp2pquic
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/vishvananda/netlink"
+)
+
+type reuseConn struct {
+	net.PacketConn
+	refCount int32 // to be used as an atomic
+}
+
+func newReuseConn(conn net.PacketConn) *reuseConn {
+	return &reuseConn{PacketConn: conn}
+}
+
+func (c *reuseConn) IncreaseCount() { atomic.AddInt32(&c.refCount, 1) }
+func (c *reuseConn) DecreaseCount() { atomic.AddInt32(&c.refCount, -1) }
+func (c *reuseConn) GetCount() int  { return int(atomic.LoadInt32(&c.refCount)) }
+
+type reuse struct {
+	mutex sync.Mutex
+
+	unicast map[string] /* IP.String() */ map[int] /* port */ *reuseConn
+	// global contains connections that are listening on 0.0.0.0 / ::
+	global map[int]*reuseConn
+}
+
+func newReuse() *reuse {
+	return &reuse{
+		unicast: make(map[string]map[int]*reuseConn),
+		global:  make(map[int]*reuseConn),
+	}
+}
+
+func (r *reuse) getSourceIPs(network string, raddr *net.UDPAddr) ([]net.IP, error) {
+	// Determine the source address that the kernel would use for this IP address.
+	// Note: This only works on Linux.
+	// On other OSes, this will return a netlink.ErrNotImplemetned.
+	routes, err := (&netlink.Handle{}).RouteGet(raddr.IP)
```

Comment:
We have to _create_ (constructor) the handle (and close it later). We should probably share one for the "reuse" instance.

Reply:
Right, I copied that code from #52. Not sure where to close it, but at least we're using the same one within one "reuse" instance.
It should also make it a bit more explicit that this is only available on Linux.